presto-main/src/main/java/com/facebook/presto/metadata/SignatureBinder.java

```diff
@@ -40,14 +40,21 @@
 {
     private final TypeManager typeManager;
     private final Signature declaredSignature;
-    private final boolean allowCoercion;
+    private final CoercionPermission coercionPermission;
```

Comment:
SignatureBinder shouldn't know anything about function selection algorithms. This abstractions should only match signatures and bind parameters. Instead of pulling here something like `coercionPermission` i would rather implement `UNKNOWN_FIRST` algorithm in the most specific function selection.

From my prespective the algorithm should look like:

1. Select all the applicable function with coercion allowed
2. Find thoose which require only UNKNOWN -> TYPE coercion
3. If ones found - try to select most specific among them
4. If ones not found - try to select most specific among all the applicable functions